### PR TITLE
Fix unsoundness in def_exc due to range in `Excluded

### PR DESCRIPTION
**Pull request by michael-schwarz**
When joining two `Definite`, we set the range of the resulting `Excluded` to be the smallest C data type that can contain both of the definite values. This gives us some precision gain (the actual value will be either one of these definite values, so any value that does not fit into this data type can be excluded). We use this additional precision in the `minimal` and `maximal` functions that are in turn called via the query system.

The problem is that the range of `Excluded` stays the same, even when doing computations: If a value that is `Excluded None byte` is e.g. incremented in a loop 500 times, we will still have `Excluded None byte` for it after the loop, which is unsound.

The attached tests uses the partitioned arrays, as they make use of this `minimal` and `maximal` via the query system, but any analysis that uses `minimal` and `maximal` is potentially unsound without the fix.

The fix modifies `join` and `lift2_inj` to correctly adapt the range, and fixes another bug in `minimal` and `maximal` where the minimal (maximal) values returned was the minimal (maximal) excluded value plus (minus) 1.

### CODE REVIEW

File: src/cdomains/intDomain.ml [hunk 1]
```diff
@@ -737,18 +739,12 @@ struct
   let min_of_range r = Option.map (fun i -> Int64.(neg @@ shift_left 1L (to_int (neg i)))) (R.minimal r)
   let maximal : t -> int64 option = function
     | `Definite x -> Integers.to_int x
-    | `Excluded (s,r) ->
-      if S.is_empty s
-      then max_of_range r
-      else Some (Int64.succ (S.max_elt s))
+    | `Excluded (s,r) -> max_of_range r
```
[the remark this answers is not in the document]
Comment (michael-schwarz) [reply src/cdomains/intDomain.ml:742]: If you have `Excluded({5,7}, short)` this will return 8.

[the remark this answers is not in the document]
Comment (michael-schwarz) [reply src/cdomains/intDomain.ml:742]: https://github.com/goblint/analyzer/blob/master/tests/regression/22-partitioned_arrays/14-with_def_exc.c
```
Expected unknown, but registered success on with_def_exc:19
    assert(j == 5); //UNKNOWN
```

**Comment by vogler** [comment]
This seems to lead to fixpoint issues:
~~~
$ ./goblint ../bench/single-thread/401.bzip2_comb.c
...
Fixpoint not reached at node "nUnused > 0" on 401.bzip2_comb.c:4328 (../bench/single-thread/401.bzip2_comb.c:4328)
 Solver computed: ...
 Difference: Map: unused =
 HoarePO: {unused[Unknown int([0,16])]} not leq {unused[Unknown int([0,8])]}.
~~~

**Comment by michael-schwarz** [comment]
Thanks! :+1: I'll have a look!

**Comment by vogler** [comment]
With aa88033b4d6f46b4d02d339477050cbcc64f5b36 reverted, it works.
Also, with `--disable ana.int.def_exc --enable ana.int.interval`, there's no issue.

**Comment by michael-schwarz** [comment]
I extracted this minimal case from the code:

```
int main(int argc , char **argv ) 
{ 
  char yy[256];
  char *ryy_j = &yy[1];

  int i =0;
  while(i < 54) { // Here
    ryy_j++;
    i++;
  }  

  return;

}
```

I'll check what's going on there tomorrow.

**Comment by michael-schwarz** [comment]
After some discussion, with Helmut we will modify the range to simply represent the range for that type. Towards this goal I'll:

- Add a range to `Definite`
- Modify so that we create an abstract value we create a `Definite` with the appropriate range

This might be less precise than what we currently do, but conceptually a lot cleaner.

**Comment by vogler** [comment]
> Modify so that we create an abstract value we create a Definite with the appropriate range

Huh?
The range is only needed for `Excluded`. Why add it do `Definite`?
The interval [a,b] represents values [0, 2^b] if a=0 and [-2^a, 2^b-1] otherwise.
If you change it to the actual range, it could only represent 32 bit integers since the interval domain is limited to it, whereas the current implementation can cover all integer types.

**Comment by michael-schwarz** [comment]
The question is what to do on a join of `Definite x` and `Definite y`.

You can either
- Construct some range such that it include both `x` and `y`. On computation, this range needs to be modified
- Set the range such that it corresponds to the entire type

What we do (before and after my patch, but broken in two different ways) is the first one.

What I got from my discussion with Helmut though is that he thinks that this conceptually the wrong thing to do, and that this range should always correspond to the type of the variable. In order to do this, and set the range so it corresponds to the type of the variable, we would need to know the type (or the range) on joining two `Definite` values. The only way I see to get this is to have the `Definite` carry this information.

Or am I missing something?

**Comment by michael-schwarz** [comment]
> The interval [a,b] represents values [0, 2^b] if a=0 and [-2^a, 2^b-1] otherwise.

I'll stick with this for sure

**Comment by vogler** [comment]
Ok.
What's wrong with approach 1? Seems like it does the right thing:
https://github.com/goblint/analyzer/blob/2c28beedf5b8b4f1faa1e70cbd7ad506c2b771ff/src/cdomains/intDomain.ml#L683-L684
The idea was that the `Excluded` range does not need to be modified on computation since overflows are taken care of by cast upon assignment.
However, I just noticed that the values in `s` need to be cast as well.
https://github.com/goblint/analyzer/blob/2c28beedf5b8b4f1faa1e70cbd7ad506c2b771ff/src/cdomains/intDomain.ml#L635
For approach 2 the problem is that you don't know the type of a literal. Only when it's assigned, it is cast to the type of the lval.

**Comment by michael-schwarz** [comment]
### Approach 1
Nothing wrong per se with approach 1, we should just agree on one approach to take.

The way I understand it, if we do approach 1, there is no way around adapting these ranges when doing computation (in the way that I patched it):

The concretization is supposed to be `𝛾(Excluded {e_1, e_2}, [a,b]) = [a=0?0:-2^a, 2^b-1] \ {e_1, e_2} `, right?

Now, consider:

```
int a = 3;
if(top) a = 7;
// a -> Excluded({0}, [0,8])
int a = a + 512
// a-> Excluded({512},[0,8})
```

In this example, there are no casts at all, and we would get  `𝛾(Excluded({512},[0,8})= [0,255]` which is clearly unsound.
Or is there something I'm missing?

### Approach 2
I don't think that this is a problem, given that constants in CIL carry their ikind:

```
type constant =
| CInt64 of int64 * ikind * string option
| ...
```
It looks like any (most?) constants that can fit into `int` are given int type, but when assigning this to an lval of a different type, it would just be the plain old `cast_to` operation, right?
So I don't think that this is a problem with approach 2 per se.

**Comment by vogler** [comment]
> Nothing wrong per se with approach 1, we should just agree on one approach to take.

1 IMO

### Approach 1

> The concretization is supposed to be `𝛾(Excluded {e_1, e_2}, [a,b]) = [a=0?0:-2^a, 2^b-1] \ {e_1, e_2} `, right?

Both bounds are different for unsigned: `𝛾(Excluded {e_1, e_2}, [a,b]) = (a=0 ? [0, 2^b] : [-2^a, 2^b-1]) \ {e_1, e_2} `,

> 
> ```
> int a = 3;
> if(top) a = 7;
> // a -> Excluded({0}, [0,8])
> int a = a + 512
> // a-> Excluded({512},[0,8})
> ```
> 
> In this example, there are no casts at all, and we would get `𝛾(Excluded({512},[0,8})= [0,255]` which is clearly unsound.
> Or is there something I'm missing?

Should be `a = a + 512` without the `int`?
Ok, so we always need to update the range on assignment but not on computation.
I.e. `(Def 3) join (Def 7)` = `Excluded ({0}, [0, 8])`, but then upon assignment it is cast to `Excluded ({0}, [-31, 31])`.
This way, we inject the least range on join for expressions, but have the full range of the type for variables.

### Approach 2
> It looks like any (most?) constants that can fit into `int` are given int type, but when assigning this to an lval of a different type, it would just be the plain old `cast_to` operation, right?

Yes, but if this is always `int`, then there's not much use for these ranges since they'd never give any precision improvement.

~~~c
byte x;
if (x != 5) // what's x?
~~~
Approach 1 = `Excluded ({5}, [0, 8])`;
Approach 2 = `Excluded ({5}, [-31, 31])`?

**Comment by michael-schwarz** [comment]
> Both bounds are different for unsigned: 𝛾(Excluded {e_1, e_2}, [a,b]) = (a=0 ? [0, 2^b] : [-2^a, 2^b-1]) \ {e_1, e_2} 
> Should be a = a + 512 without the int?

True, I shouldn't comment on issues early in the morning.

# Approach 1

> Ok, so we always need to update the range on assignment but not on computation.

What about code such as

```
int a = 3;
if (top) a= 7;
if(a + 512 == 519) {

}
```
There is no assignment to a after the join, so presumably we would have `Excluded({512},[0,8])` for `a+512` according to this scheme right? So `Excluded({512},[0,8]) == 519` would evaluate to `Definite 0` here? Which is unsound.


## Approach 2

```
byte x;
if (x != 5) // what's x?
```
Even in Approach 2, this would be `meet(Excluded {} [0,8], Excluded {5} [-63, 63]) = Excluded {5} [0,8]`, right?
https://github.com/goblint/analyzer/blob/9376fc0dc44d8ba599aea8ce40724e0fd9c47776/src/analyses/base.ml#L1374-L1376

This would be handled by this `invariant` thing via

**Comment by vogler** [comment]
Yes, tricky. We only want to increase the range when an overflow would occur and it's not the full range yet.
So, `cast_to` would have to be called after every operation and check.
~~~c
int a = 3; // Def 3
if (top)
  a= 7; // Def 7
// (Def 3) join (Def 7) = Exc {0} [0,8]
if(a + 512 == 519) { // Exc {0} [0,8] + Def 512 == Def 519
...
~~~
becomes `cast_to Int (Exc {512} [0,8]) == Def 519` ... `Exc {512} [-31, 31] == Def 519`.
Alternatively, there could be some threshold widening for the range instead of going to the full range in `cast_to`.

**Comment by michael-schwarz** [comment]
I made some progress with the second approach and it seems that we do not lose too much precision (at least going by the regression tests) and the logic is a lot simpler IMO.
I'll finish this and commit it in the next few days (whenever some students don't show up), and then we can discuss further based on that :smiley:

**Comment by vogler** [comment]
Can you undelete this PR's branch and rebase your commits onto it? Like this it's unconnected and reviewing doesn't work.

**Comment by michael-schwarz** [comment]
I'll do so. This is still not finished though, there are still some issues with the arrays if I enable `def_exc` for `22` and `23`. So I thinks it makes more sense for you to review this once it's all fixed.

**Pull request merged**